File: ujin/2주차_수학/1182.py
def DFS(L,sum):
    global count
    if L == N:
        if sum == S:
            count += 1
            return
        return
    else:
        DFS(L+1,sum + number[L])
        DFS(L+1,sum)

N, S = map(int, input().split())
number = list(map(int, input().split()))
count = 0
DFS(0,0)
if S==0:
    print(count-1)
else:
    print(count)

'''
재귀를 통해 나올 수 있는 모든 부분수열에 대해 검사했다. (브루트포스 알고리즘)
구해진 부분수열의 합이 S와 같으면 count값을 증가시키도록 했다.
'''

File: minto/2주차_수학/1182.js
let input = require('fs').readFileSync('/dev/stdin').toString().split('\n');

const n = +input[0].split(' ')[0];
const s = +input[0].split(' ')[1];

const inputArray = input[1]
  .split(' ')
  .map((item) => +item)
  .sort((a, b) => a - b);

let result = 0;

function getRecursion(k, sum) {
  if (k === n) {
    if (sum === s) {
      result += 1;
    }
    return;
  }
  getRecursion(k + 1, sum + inputArray[k]);
  getRecursion(k + 1, sum);
}

getRecursion(0, 0);

if (s === 0 && result >= 1) result -= 1;
console.log(result);

/*
  브루트포스 알고리즘을 생각했다.
  왜냐하면 N의 갯수가 20개밖에 되지 않아서 모든 경우의수를 검색해도 되겠다라는 생각을 먼저했고 

  정답이 되는 경우의 수의 중복을 검사할 방법이 딱히 없는거 같에서 재귀를 돌면서 맞는 경우만 +1을 하도록 구현했다.
*/
